### PR TITLE
fix: `json.loads` trailing comma

### PR DESCRIPTION
**Pull request by TheSecEng**
**Issue**
When syncing settings, some sublime-settings file contain trailing comma's. The `json` package doesn't handle trailing comma's very well.

**Solution**

1. Removed `json` library
2. Implemented `sublime.encode_value` and `sublime.decode_value`

### CODE REVIEW

**Comment by mfuentesg** [comment]
Hello, please, can you rebase your branch with the latest changes.

File: sync_settings/commands/download.py [hunk 1]
```diff
@@ -1,23 +1,26 @@
...
+from ..libs import path, settings
 from ..libs.gist import Gist
 from ..libs.logger import logger
 from ..thread_progress import ThreadProgress
+from . import decorators
 
 
 class SyncSettingsDownloadCommand(sublime_plugin.WindowCommand):
-    temp_folder = path.join(os.path.expanduser('~'), '.sync_settings', 'temp')
+    temp_folder = path.join(os.path.expanduser("~"), ".sync_settings", "temp")
```
Comment (mfuentesg) [inline sync_settings/commands/download.py:18]: keep single quotes for everything please.

**Comment by mfuentesg** [comment]
It's very difficult to follow the changes due to all "quotes" changes.

Please, just keep related changes and keep using single quotes for everything.

**Comment by TheSecEng** [comment]
Sure, and sorry, that was my LSP correcting the quotes. I’ll rebase and revert quote changes tonight

File: sync_settings/libs/gist.py [hunk 2]
```diff
@@ -63,14 +65,22 @@ def make_uri(endpoint=''):
     def create(self, data):
         if not isinstance(data, dict) or not len(data):
             raise ValueError('Gist can`t be created without data')
-        return self.__do_request('post', self.make_uri(), data=json.dumps(data)).json()
+        return self.__do_request(
+             'post',
+             self.make_uri(),
+             data=sublime.encode_value(data, True).encode('utf-8').decode('latin-1'),
```
Comment (TheSecEng) [inline sync_settings/libs/gist.py:71]: @mfuentesg with this line, I've done a encode -> decode due to the fact that some settings and files can cause detection of latin-1 encoding. This resolves that, it doesn't impact anything else.

Comment (mfuentesg) [reply sync_settings/libs/gist.py:71]: can you create a unit test for this specific case?

File: sync_settings/libs/gist.py [hunk 3]
```diff
@@ -63,14 +65,22 @@ def make_uri(endpoint=''):
...
     @auth
     @with_gid
     def update(self, gid, data):
         if not isinstance(data, dict) or not len(data):
             raise ValueError('Gist can`t be updated without data')
-        return self.__do_request('patch', self.make_uri(gid), data=json.dumps(data)).json()
+        return self.__do_request(
+             'patch',
+             self.make_uri(gid),
+             data=sublime.encode_value(data, True).encode('utf-8').decode('latin-1'),
```
Comment (mfuentesg) [inline sync_settings/libs/gist.py:82]: same here

File: sync_settings/sync_version.py [hunk 4]
```diff
@@ -37,7 +38,7 @@ def get_remote_version():
 
 def update_config_file(info):
     with open(file_path, 'w') as f:
-        json.dump(info, f)
+        f.write(sublime.encode_value(info, True))
```
Comment (mfuentesg) [inline sync_settings/sync_version.py:41]: add unit tests to ensure that all changes are solving the issues that you are mentioning.

**Comment by mfuentesg** [comment]
In general LGTM, but you need to ensure that your changes are working properly.

**Comment by TheSecEng** [comment]
I'll work on the test case tonight

File: sync_settings/sync_version.py (hunk 4 above)
Comment (TheSecEng) [reply sync_settings/sync_version.py:41]: 👍

**Comment by TheSecEng** [comment]
@mfuentesg I am having trouble creating a test case for this, I can recreate the issue with my files locally every time, however creating a test case isn't appearing as easy.

For the time being, I will revert a few changes until I can get a test case working.

File: sync_settings/libs/gist.py (hunk 3 above)
Comment (TheSecEng) [reply sync_settings/libs/gist.py:82]: @mfuentesg turns out this isn't a proper solution as the issue causing this is when reading binary files such as a image file. Which `json.dumps` also handles incorrectly.
I am going to revert most changes here except that which solves the trailing comma and open a new PR for a solution for handling binary data.

Comment (TheSecEng) [reply sync_settings/libs/gist.py]: reverted changes due to larger PR coming

File: sync_settings/libs/gist.py (hunk 2 above)
Comment (TheSecEng) [reply sync_settings/libs/gist.py]: reverted changes due to larger PR coming

**Comment by TheSecEng** [comment]
@mfuentesg I have never written a test case. Can you validate this one?

**Comment by TheSecEng** [comment]
@mfuentesg going to re-write this with a more clean implementation

**Pull request closed without merging**